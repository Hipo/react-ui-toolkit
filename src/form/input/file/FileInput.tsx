import React from "react";
import classNames from "classnames";

import Spinner from "../../../spinner/Spinner";

// SCSS import is moved here to be able to override Spinner styles without nesting
import "./_file-input.scss";

export interface FileInputProps {
  htmlFor: string;
  name: string;
  children: React.ReactNode;
  onChange: React.ReactEventHandler<HTMLInputElement>;
  testid?: string;
  isDisabled?: boolean;
  isPending?: boolean;
  customSpinner?: React.ReactNode;
  customClassName?: string;
  customLabelClassName?: string;
  acceptedFileTypes?: string;
  labelRef?: React.RefObject<HTMLLabelElement>;
}

function FileInput({
  onChange,
  children,
  testid,
  name,
  htmlFor,
  acceptedFileTypes = "image/png, image/jpeg, .pdf",
  customSpinner,
  customClassName,
  customLabelClassName,
  isPending,
  isDisabled,
  labelRef
}: FileInputProps) {
  const containerClassName = classNames("file-input__container", customClassName);
  const isInputDisabled = isPending || isDisabled;
  const labelClassName = classNames("file-input__label", customLabelClassName, {
    "file-input__label--is-disabled": isInputDisabled
  });
  const displaySpinner = customSpinner || (
    <Spinner customClassName={"file-input__spinner"} />
  );

  return (
    <div className={containerClassName}>
      <input
        data-testid={testid}
        type={"file"}
        value={""}
        name={name}
        className={"file-input"}
        onChange={onChange}
        id={htmlFor}
        disabled={isInputDisabled}
        accept={acceptedFileTypes}
      />

      <label
        ref={labelRef}
        htmlFor={htmlFor}
        className={labelClassName}
        data-testid={`${testid}.label`}>
        {children}

        {isPending && displaySpinner}
      </label>
    </div>
  );
}

export default FileInput;
