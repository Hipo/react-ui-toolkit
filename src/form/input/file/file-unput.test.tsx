import React from "react";
import {render, screen} from "@testing-library/react";
import "@testing-library/jest-dom";
import {create} from "react-test-renderer";

import {testA11y} from "../../../core/utils/test/testUtils";
import FileInput, {FileInputProps} from "./FileInput";

describe("<FileInput />", () => {
  const defaultFileInputProps: FileInputProps = {
    testid: "file-input",
    name: "file-input",
    htmlFor: "file-input",
    onChange: jest.fn(),
    children: "Upload File"
  };

  it("should render correctly", () => {
    render(<FileInput {...defaultFileInputProps} />);
  });

  it("should match snapshot", () => {
    const tree = create(<FileInput {...defaultFileInputProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should pass a11y test", async () => {
    const {container} = render(<FileInput {...defaultFileInputProps} />);

    await testA11y(container);
  });

  it("should render children correctly", () => {
    render(<FileInput {...defaultFileInputProps} />);

    expect(screen.getByText("Upload File")).toHaveTextContent("Upload File");
  });

  it("should display custom spinner correctly", () => {
    const customSpinnerContent = <p data-testid={"custom-spinner"}>{"Loading..."}</p>;

    render(
      <FileInput
        isPending={true}
        customSpinner={customSpinnerContent}
        {...defaultFileInputProps}
      />
    );

    const customSpinner = screen.getByText("Loading...");

    expect(screen.getByText("Upload File")).toContainElement(customSpinner);
  });

  it("should add disabled attribute to file input and file-input__label--is-disabled class to file input label when isDisabled is true", () => {
    render(<FileInput isDisabled={true} {...defaultFileInputProps} />);

    expect(screen.getByLabelText("Upload File")).toBeDisabled();
    expect(screen.getByText("Upload File")).toHaveClass("file-input__label--is-disabled");
  });

  it("should add disabled attribute to file input and file-input__label--is-disabled class to file input label when isPending is true", () => {
    render(<FileInput isPending={true} {...defaultFileInputProps} />);

    expect(screen.getByLabelText("Upload File")).toBeDisabled();
    expect(screen.getByText("Upload File")).toHaveClass("file-input__label--is-disabled");
  });
});
