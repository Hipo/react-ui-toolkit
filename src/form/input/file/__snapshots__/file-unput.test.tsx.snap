// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileInput /> should match snapshot 1`] = `
<div
  className="file-input__container"
>
  <input
    accept="image/png, image/jpeg, .pdf"
    className="file-input"
    data-testid="file-input"
    id="file-input"
    name="file-input"
    onChange={[MockFunction]}
    type="file"
    value=""
  />
  <label
    className="file-input__label"
    data-testid="file-input.label"
    htmlFor="file-input"
  >
    Upload File
  </label>
</div>
`;
